optuna/samplers/nsgaiii.py

```diff
@@ -0,0 +1,643 @@
+from optuna.study import Study
+from optuna.study._multi_objective import _dominates
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+# Define key names of `Trial.system_attrs`.
+_GENERATION_KEY = "nsga3:generation"
+_POPULATION_CACHE_KEY_PREFIX = "nsga3:population"
+# Define a coefficient for scaling intervals, used in _filter_inf() to replace +-inf
+_COEF = 3
+
+
+class NSGAIIISampler(BaseSampler):
+    """Multi-objective sampler using the NSGA-III algorithm.
+
+    NSGA-III stands for "Nondominated Sorting Genetic Algorithm III",
+    which is a modified version of NSGA-II for many objective optimization problem.
+
+    For further information about NSGA-III, please refer to the following papers:
+
+    - `An Evolutionary Many-Objective Optimization Algorithm Using Reference-Point-Based
+      Nondominated Sorting Approach, Part I: Solving Problems With Box Constraints
+      <https://doi.org/10.1109/TEVC.2013.2281535>`_
+    - `An Evolutionary Many-Objective Optimization Algorithm Using Reference-Point-Based
+      Nondominated Sorting Approach, Part II: Handling Constraints and Extending to an Adaptive
+      Approach
+      <https://doi.org/10.1109/TEVC.2013.2281534>`_
+
+    Args:
+        reference_points:
+            A 2 dimension ``numpy.ndarray`` with objective dimension columns. Represents
+            a list of reference points which is used to determine who to survive.
+            After non-dominated sort, who out of borderline front are going to survived is
+            determined according to how sparse the closest reference point of each individual is.
+            In the default setting the algorithm uses `uniformly` spread points to diversify the
+            result. It is also possible to reflect your `preferences` by giving an arbitrary set of
+            `target` points since the algorithm prioritizes individuals around reference points.
+
+        dividing_parameter:
+            An parameter to determine the density of default reference points. This parameter
+            determines how many divisions are made between reference points on each axis. The
+            smaller this value is, the less reference points you have. The default value is 3.
+            Note that this parameter is not used when reference_points is not None.
+
+    .. note::
+        Other parameters than reference_points and dividing_parameter are the same as
+        :class:`~optuna.samplers.nsgaii.NSGAIISampler`.
+
+    """
+
+    def __init__(
+        self,
+        population_size: int = 50,
+        reference_points: np.ndarray | None = None,
+        dividing_parameter: int = 3,
+        mutation_prob: float | None = None,
+        crossover: BaseCrossover | None = None,
+        crossover_prob: float = 0.9,
+        swapping_prob: float = 0.5,
+        seed: int | None = None,
+        constraints_func: Callable[[FrozenTrial], Sequence[float]] | None = None,
```

Comment:
How about following the argument order of `NSGAIISampler.__init__`?

Reply:
I see. I sorted the argument as in `NSGAIISampler.__init__()`